File: packages/agent-server/src/router/messages.ts
import fs from "fs";
import { router, publicProcedure } from "./trpc.js";
import {
  sendParamsSchema,
  listMessagesParamsSchema,
  getMediaParamsSchema,
} from "@thisnick/agent-wechat-shared";
import type { SendResult, Message, MediaResult } from "@thisnick/agent-wechat-shared";
import { getStoredKeys, getImageKeys, storeSingleKey } from "../lib/wechat-keys.js";
import { listMessagesFromWechatDb } from "../lib/wechat-messages.js";
import { getMessageMedia } from "../lib/wechat-media.js";
import { getDb } from "../db/index.js";
import { createContext } from "../context/index.js";
import { createExecution, runExecution } from "../execution/index.js";
import { sendMessagePlan } from "../plans/index.js";

export const messagesRouter = router({
  /**
   * List messages for a specific chat
   */
  list: publicProcedure
    .input(listMessagesParamsSchema)
    .query(async ({ input, ctx }): Promise<Message[]> => {
      const session = ctx.session;
      if (!session?.loggedInUser) return [];

      const keys = getStoredKeys(ctx.db, session.id, session.loggedInUser);
      if (!keys["message_0.db"]) return [];

      return listMessagesFromWechatDb(session.loggedInUser, keys, input.chatId, input.limit, input.offset);
    }),

  /**
   * Get media attachment for a message (image thumbnail, emoji, or voice)
   */
  media: publicProcedure
    .input(getMediaParamsSchema)
    .query(async ({ input, ctx }): Promise<MediaResult> => {
      const session = ctx.session;
      if (!session?.loggedInUser) return { type: "unsupported", format: "", filename: "" };

      const accountDir = session.loggedInUser;
      const keys = getStoredKeys(ctx.db, session.id, accountDir);
      const imageKeys = getImageKeys(ctx.db, session.id, accountDir);
      return getMessageMedia(accountDir, keys, input.chatId, input.localId, imageKeys,
        (xorByte) => storeSingleKey(ctx.db, session.id, accountDir, "_image_xor", xorByte.toString(16).padStart(2, "0")),
      );
    }),

  /**
   * Send a text message, image, or file via FSM plan
   */
  send: publicProcedure
    .input(sendParamsSchema)
    .mutation(async ({ input, ctx }): Promise<SendResult> => {
      if (!input.text && !input.image && !input.file) {
        return { success: false, error: "No text, image, or file provided" };
      }

      const session = ctx.session;
      if (!session) {
        return { success: false, error: "No session available" };
      }
      if (!session.loggedInUser) {
        return { success: false, error: "NOT_LOGGED_IN" };
      }

      // Decode base64 image to temp file (if provided)
      let imagePath: string | undefined;
      let imageMime: string | undefined;
      if (input.image) {
        const ext = input.image.mimeType === "image/jpeg" ? ".jpg" :
                    input.image.mimeType === "image/gif" ? ".gif" : ".png";
        imagePath = `/tmp/send_image_${Date.now()}${ext}`;
        fs.writeFileSync(imagePath, Buffer.from(input.image.data, "base64"));
        imageMime = input.image.mimeType;
      }

      // Decode base64 file to temp file (if provided)
      let filePath: string | undefined;
      if (input.file) {
        filePath = `/tmp/send_file_${Date.now()}_${input.file.filename}`;
        fs.writeFileSync(filePath, Buffer.from(input.file.data, "base64"));
      }

      try {
        const db = getDb();
        const context = await createContext(session, db);

        const execution = createExecution(
          sendMessagePlan,
          {
            chatId: input.chatId,
            message: input.text,
            imagePath,
            imageMime,
            filePath,
          },
          context,
          {
            emit: () => {},
            abortSignal: ctx.abortSignal,
          }
        );

        const result = await runExecution(execution);
        return { success: result.success, error: result.error };
      } finally {
        // Clean up temp files
        if (imagePath) {
          try { fs.unlinkSync(imagePath); } catch {}
        }
        if (filePath) {
          try { fs.unlinkSync(filePath); } catch {}
        }
      }
    }),
});

File: packages/agent-server-rust/src/router/messages.rs
use axum::{
    extract::{Path, Query},
    Json,
};
use serde::Deserialize;

use crate::db::get_db;
use crate::ia::types::{MediaResult, Message, SendResult};
use crate::tools::wechat_keys::{get_stored_keys, get_image_keys};
use crate::tools::wechat_media::get_message_media;
use crate::tools::wechat_messages;
use crate::sessions::manager::get_session;

#[derive(Deserialize)]
pub struct ListParams {
    #[serde(default = "default_limit")]
    limit: i64,
    #[serde(default)]
    offset: i64,
}

fn default_limit() -> i64 {
    50
}

pub async fn list_messages(
    Path(chat_id): Path<String>,
    Query(params): Query<ListParams>,
) -> Json<Vec<Message>> {
    let session = match get_session("default") {
        Some(s) => s,
        None => return Json(Vec::new()),
    };
    let logged_in_user = match &session.logged_in_user {
        Some(u) => u.clone(),
        None => return Json(Vec::new()),
    };

    let db = get_db();
    let keys = get_stored_keys(&db, &session.id, &logged_in_user);
    if !keys.contains_key("message_0.db") {
        return Json(Vec::new());
    }

    Json(wechat_messages::list_messages(
        &logged_in_user,
        &keys,
        &chat_id,
        params.limit,
        params.offset,
    ))
}

pub async fn get_media(Path((chat_id, local_id)): Path<(String, i64)>) -> Json<MediaResult> {
    let session = match get_session("default") {
        Some(s) => s,
        None => {
            return Json(MediaResult {
                media_type: "unsupported".to_string(),
                data: None,
                url: None,
                format: String::new(),
                filename: String::new(),
            })
        }
    };
    let logged_in_user = match &session.logged_in_user {
        Some(u) => u.clone(),
        None => {
            return Json(MediaResult {
                media_type: "unsupported".to_string(),
                data: None,
                url: None,
                format: String::new(),
                filename: String::new(),
            })
        }
    };

    let db = get_db();
    let keys = get_stored_keys(&db, &session.id, &logged_in_user);
    let image_keys = get_image_keys(&db, &session.id, &logged_in_user);

    Json(get_message_media(
        &logged_in_user,
        &keys,
        &chat_id,
        local_id,
        image_keys,
    ))
}

#[derive(Deserialize)]
pub struct SendParams {
    #[serde(rename = "chatId")]
    chat_id: String,
    text: Option<String>,
    image: Option<ImageInput>,
    file: Option<FileInput>,
}

#[derive(Deserialize)]
pub struct ImageInput {
    data: String,
    #[serde(rename = "mimeType")]
    mime_type: String,
}

#[derive(Deserialize)]
pub struct FileInput {
    data: String,
    filename: String,
}

pub async fn send_message(Json(input): Json<SendParams>) -> Json<SendResult> {
    if input.text.is_none() && input.image.is_none() && input.file.is_none() {
        return Json(SendResult {
            success: false,
            error: Some("No text, image, or file provided".to_string()),
        });
    }

    // TODO: Run send-message FSM plan
    Json(SendResult {
        success: false,
        error: Some("Not yet implemented".to_string()),
    })
}
